src/core/exec-invoke.c

```diff
@@ -4212,6 +4212,166 @@ static bool exec_context_need_unprivileged_private_users(
                !strv_isempty(context->no_exec_paths);
 }
 
+static bool exec_namespace_is_delegated(const ExecContext *context, const ExecParameters *params, unsigned long namespace) {
+        /* If we need unprivileged private users, we've already unshared a user namespace by the time we call
+         * setup_namespaces() for the first time so let's make sure we do all other namespace unsharing in
+         * the first call to setup_namespaces() by returning false here. */
+        if (exec_context_need_unprivileged_private_users(context, params))
+                return false;
+
+        if (context->private_users == PRIVATE_USERS_NO)
+                return false;
+
+        if (context->delegate_namespaces == NAMESPACE_FLAGS_INITIAL)
+                return false;
+
+        return FLAGS_SET(context->delegate_namespaces, namespace);
+}
+
+static int setup_namespaces(
```

Comment:
maybe call this "setup_delegated_namespaces", to emphasize that this won't do userns but only those delegated to userns